controllers/kubeadmconfig_controller.go

```diff
@@ -263,6 +267,83 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
+	if config.Spec.JoinConfiguration.Discovery.File != nil {
+		return nil
+	}
+
+	// otherwise it is necessary to ensure token discovery is properly configured
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
+	}
+
+	// if BootstrapToken already contains an APIServerEndpoint, respect it; otherwise inject the APIServerEndpoint endpoint defined in cluster status
+	//TODO(fp) might be we want to validate user provided APIServerEndpoint and warn/error if it doesn't match the api endpoint defined at cluster level
+	apiServerEndpoint := config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint
+	if apiServerEndpoint == "" {
+		if len(cluster.Status.APIEndpoints) == 0 {
+			return errors.New("Control plane already exists for the cluster but cluster.Status.APIEndpoints are not set")
+		}
+
+		// NB. CABPK only uses the first APIServerEndpoint defined in cluster status if there are multiple defined.
+		apiServerEndpoint = fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port)
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint = apiServerEndpoint
+		log.Info("Altering JoinConfiguration.Discovery.BootstrapToken", "APIServerEndpoint", apiServerEndpoint)
+	}
+
+	// if BootstrapToken already contains a token, respect it; otherwise create a new bootstrap token for the node to join
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token == "" {
+		// if there is no secretInterfaceFactory in place
+		if r.secretInterfaceFactory == nil {
+			// create the default secretInterfaceFactory function, that connects to the given remote cluster
+			// and then gets the secretInterface client for the kube-system namespace
+			r.secretInterfaceFactory = func(cluster *capiv1alpha2.Cluster) (corev1.SecretInterface, error) {
+				remoteClient, err := capiremote.NewClusterClient(r.Client, cluster)
+				if err != nil {
+					return nil, err
+				}
+
+				corev1Client, err := remoteClient.CoreV1()
+				if err != nil {
+					return nil, err
+				}
+
+				return corev1Client.Secrets(metav1.NamespaceSystem), nil
+			}
+		}
+
+		// gets the remote secret interface client for the current cluster
+		remoteSecretInterfaceClient, err := r.secretInterfaceFactory(cluster)
+		if err != nil {
+			return err
+		}
+
+		token, err := createToken(remoteSecretInterfaceClient)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create new bootstrap token")
+		}
+
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token = token
+		log.Info("Altering JoinConfiguration.Discovery.BootstrapToken", "Token", token)
+	}
+
+	// if BootstrapToken already contains a CACertHashes or UnsafeSkipCAVerification, respect it; otherwise set for UnsafeSkipCAVerification
+	// TODO: set CACertHashes instead of UnsafeSkipCAVerification
```

Comment:
Please file an issue for this as this must make it into the current milestone.

Reply:
https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/issues/123